File: LeetCode 702 - Search in a Sorted Array of Unknown Size.java
// Time Complexity: Since we use binary search methods here, 
// so the time complexity will be O(logn).
/* --------------------------------------------------------*/
// Space Complexity: the space complexity will be O(1).

class Solution {
    public int search(ArrayReader reader, int target) {
        int low = 0, high = 1;
        while(low <= high){
            int mid = low + (high - low)/2;
            // pay attention to the logic:
            // we need firstly check whether the high is greater than target,
            // so that we can confirm the interval or check whether we need move forward.
            if(reader.get(high) < target){
                low = high + 1;
                high = 2 * low;
            }else{
                if(reader.get(mid) == target){
                    return mid;
                }else if(reader.get(mid) < target){
                        low = mid + 1;

                }else{
                        high = mid - 1;
                }
            }
        }
        return -1;
    }
}

File: Leetcode-702_Search in a Sorted Array of Unknown Size.cpp
// Time Complexity : O(Log(Index))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * // This is the ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * class ArrayReader {
 *   public:
 *     int get(int index);
 * };
 */

class Solution 
{
    public:
        int search(const ArrayReader& reader, int target) 
        {
            int low = 0;
            int high = 1;
            
			if(reader.get(low)==target)
                return low;
			
            // get search space first
            while(reader.get(high)<target)
            {
                low = high;
                high = high<<1;
            }
            
			
			// do binary search in above search space
            while(low <=high)
            {
                int mid = low+ (high-low)/2;
                
                int val = reader.get(mid);
                if(val == target)
                    return mid;
                
                if(val>target)
                    high = mid-1;
                
                else
                    low = mid+1;
            }
            
            return -1;
        }
};